walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -34,55 +36,109 @@
-
-    if (request.getAttendees().isEmpty()) {
-      return optionalFreeTimes;
-    } else if (request.getOptionalAttendees().isEmpty()) {
+    if (optionalBusyTimes.isEmpty()) {
       return mandatoryFreeTimes;
     }
 
-    Collection<TimeRange> bothFreeTimes = findIntersectionFreeTimes(mandatoryFreeTimes, optionalFreeTimes, request.getDuration());
-    return bothFreeTimes.isEmpty() ? mandatoryFreeTimes : bothFreeTimes;
+    // At index i of optionalFreeTimes, i + 1 optional attendees can attend any of the given time ranges.
+    List<List<TimeRange>> optionalFreeTimes = makeOptionalFreeTimesList(optionalBusyTimes, request.getDuration());
+    
+    List<TimeRange> bothFreeTimes;
+    // Iterate back to front through the list so that time ranges with the max number of attendees are returned.
+    for (int i = optionalFreeTimes.size() - 1; i >= 0; i--) {
+      if (request.getAttendees().isEmpty() && !optionalFreeTimes.get(i).isEmpty()) {
+        Collections.sort(optionalFreeTimes.get(i), TimeRange.ORDER_BY_START);
+        return optionalFreeTimes.get(i);
+      }
+      bothFreeTimes = findIntersectionFreeTimes(mandatoryFreeTimes, optionalFreeTimes.get(i), request.getDuration());
+      if (!bothFreeTimes.isEmpty()) {
+        Collections.sort(bothFreeTimes, TimeRange.ORDER_BY_START);
+        return bothFreeTimes;
+      }
+    }
+    return mandatoryFreeTimes;
   }
 
   private static boolean hasCommonAttendees(Collection<String> eventAttendees, Collection<String> requestAttendees) {
     return !Collections.disjoint(eventAttendees, requestAttendees);
   }
 
+  private static Collection<String> getCommonAttendees(Collection<String> eventAttendees, Collection<String> requestAttendees) {
+    Collection<String> commonAttendees = new HashSet<>();
+    for (String person : eventAttendees) {
+      if (requestAttendees.contains(person)) {
+        commonAttendees.add(person);
+      }
+    }
+    return commonAttendees;
+  }
+
+  // Returns a nested list where the index of each list represents that index + 1 optional attendees can attend those times.
+  private static List<List<TimeRange>> makeOptionalFreeTimesList(
+      Map<String, Set<TimeRange>> optionalBusyTimesMap, long meetingDuration) {
+    List<List<TimeRange>> optionalFreeTimes = new ArrayList<>();
+    int numOptionalAttendees = optionalBusyTimesMap.size();
+    for (int i = 0; i < numOptionalAttendees; i++) {
+      optionalFreeTimes.add(new ArrayList<TimeRange>());
+    }
+    for (Map.Entry entry : optionalBusyTimesMap.entrySet()) {
+      List<TimeRange> freeTimes = findFreeTimes((Collection<TimeRange>) entry.getValue(), meetingDuration);
+      for (TimeRange freeRange : freeTimes) {
+        for (int i = numOptionalAttendees - 1; i >= 0; i--) {
+          for (TimeRange existingRange : optionalFreeTimes.get(i)) {
+            TimeRange intersection = freeRange.intersection(existingRange);
+            if (i < numOptionalAttendees - 1 && 
```

Comment:
also, not exactly a "solution" to lots of nested loops, but sometimes it can be clearer to read parts of it by extracting out logic in loops to even more smaller methods 
for (TimeRange freeRange : freeTimes):
    addToOptionalFreeTimes(freeRange, optionalFreeTimes);

.. etc